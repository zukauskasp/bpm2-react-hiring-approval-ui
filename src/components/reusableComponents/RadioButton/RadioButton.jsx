import React from "react";
import Tooltip from "../Tooltip/Tooltip";
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';

import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormControl from '@material-ui/core/FormControl';
import FormLabel from '@material-ui/core/FormLabel';


const RadioButton = ({ label, items, onChange, selectedItem, errorMessage }) => {

    const classes = useStyles();

    return (
        <FormControl className={classes.formControl}>
            {label ? <FormLabel className={classes.customLabel}>{label}</FormLabel> : null}
            <RadioGroup  >
                {items && items.map((option, index) => (
                    <FormControlLabel
                        key={index} value={option.value}
                        control={<StyledRadio />}
                        label={option.name}
                        disabled={!!option.disabled ? true : false}
                        className={`${classes.customLabel} ${classes.radioLabel}`}
                        onChange={onChange}
                        checked={selectedItem && selectedItem === option.value ? true : false}
                    />
                ))}
            </RadioGroup>
            {errorMessage ? <Tooltip text={errorMessage} messageType='error' /> : null}
        </FormControl>
    );

}

RadioButton.propTypes = {
    errorMessage: PropTypes.string,
    label: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string,
        value: PropTypes.string,
        disabled: PropTypes.bool
    })).isRequired,
    selectedItem: PropTypes.string.isRequired
}

RadioButton.defaultProps = {
    errorMessage: "",
    label: ""
}


export default RadioButton;

const useStyles = makeStyles(theme => ({
    formControl: {
        marginTop: '24px'
    },
    customLabel: {
        fontFamily: 'Danske Text v2',
        color: "#003755",
        marginBottom: "8px"
    },
    radioLabel: {
        marginBottom: 0
    }
}));

function StyledRadio(props) {
    const classes = useStyles2();

    return (
        <Radio
            className={classes.root}

            color="default"
            checkedIcon={<span className={` ${classes.checkedIcon}`} />}
            icon={<span className={classes.icon} />}
            {...props}
            disableRipple
        />
    );
}

const useStyles2 = makeStyles({
    root: {
        '&:hover': {
            backgroundColor: 'transparent',
        },
    },
    icon: {
        borderRadius: '50%',
        width: 16,
        height: 16,
        border: '1px solid rgb(0, 159, 218)',
        backgroundColor: '#f5f8fa',
        backgroundImage: 'linear-gradient(180deg,hsla(0,0%,100%,.8),hsla(0,0%,100%,0))',
        '$root.Mui-focusVisible &': {
            outline: '2px auto rgba(19,124,189,.6)',
            outlineOffset: 2,
        },
        'input:hover ~ &': {
            backgroundColor: '#ebf1f5',
        },
        'input:disabled ~ &': {
            boxShadow: 'none',
            background: 'rgba(206,217,224,.5)',
            border: '1px solid rgb(206,217,224)'
        },
    },
    checkedIcon: {
        backgroundColor: 'rgb(0, 159, 218)',
        backgroundImage: 'linear-gradient(180deg,hsla(0,0%,100%,.1),hsla(0,0%,100%,0))',
        borderRadius: '50%',
        '&:before': {
            display: 'block',
            width: 16,
            height: 16,
            backgroundImage: 'radial-gradient(#fff,#fff 28%,transparent 32%)',
            content: '""',
        },
        'input:hover ~ &': {
            backgroundColor: 'rgb(0, 159, 218)',
        },
        'input:disabled ~ &': {
            boxShadow: 'none',
            background: 'rgba(206,217,224,.5)',
            border: '1px solid rgb(206,217,224)'
        },
    }
});
