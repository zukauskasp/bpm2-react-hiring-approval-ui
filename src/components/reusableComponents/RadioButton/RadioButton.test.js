import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import RadioButton from './RadioButton';

describe('RadioButton', () => {

    let wrapper;
    let mount;
    let shallow;

    beforeAll(() => {
        shallow = createShallow();
        mount = createMount();
    });

    const list = [
        { name: "one", value: "1" },
        { name: "two", value: "2" },
        { name: "three", value: "3", disabled: true },
    ]

    const props = {
        items: list,
        label: "Select number",
        onChange: () => { },
        selectedItem: "",
        errorMessage: "Error message"
    }

    beforeEach(() => {
        wrapper = shallow(<RadioButton {...props} />);
    })

    afterAll(() => {
        mount.cleanUp();
      });

    it('expect to match latest snapshot', () => {
        expect(shallow(<RadioButton {...props} />).debug()).toMatchSnapshot();
    });

    it('expect to render RadioButton', (() => {
        expect(wrapper.length).toEqual(1);
    }));

    it('expect to render correct props', () => {
        expect(mount(<RadioButton {...props} />).props()).toEqual(props);
    });

    it('expect to render default props', () => {
        const mounted = mount(<RadioButton items={props.items} onChange={props.onChange} selectedItem={props.selectedItem} />);

        expect(mounted.props().errorMessage).toEqual("");
        expect(mounted.props().label).toEqual("");
    });

    it('expect to find correct radio buttons', () => {
        expect(wrapper.childAt(1).children().length).toEqual(props.items.length);
        expect(wrapper.childAt(1).find({ disabled: true }).length).toEqual(1);
    });
})