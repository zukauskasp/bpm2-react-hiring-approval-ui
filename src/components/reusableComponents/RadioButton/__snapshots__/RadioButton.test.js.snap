// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton expect to match latest snapshot 1`] = `
"<WithStyles(ForwardRef(FormControl)) className=\\"makeStyles-formControl-1\\">
  <WithStyles(ForwardRef(FormLabel)) className=\\"makeStyles-customLabel-2\\">
    Select number
  </WithStyles(ForwardRef(FormLabel))>
  <ForwardRef(RadioGroup)>
    <WithStyles(ForwardRef(FormControlLabel)) value=\\"1\\" control={{...}} label=\\"one\\" disabled={false} className=\\"makeStyles-customLabel-2 makeStyles-radioLabel-3\\" onChange={[Function: onChange]} checked={false} />
    <WithStyles(ForwardRef(FormControlLabel)) value=\\"2\\" control={{...}} label=\\"two\\" disabled={false} className=\\"makeStyles-customLabel-2 makeStyles-radioLabel-3\\" onChange={[Function: onChange]} checked={false} />
    <WithStyles(ForwardRef(FormControlLabel)) value=\\"3\\" control={{...}} label=\\"three\\" disabled={true} className=\\"makeStyles-customLabel-2 makeStyles-radioLabel-3\\" onChange={[Function: onChange]} checked={false} />
  </ForwardRef(RadioGroup)>
  <Message text=\\"Error message\\" messageType=\\"error\\" position=\\"top-left\\" />
</WithStyles(ForwardRef(FormControl))>"
`;
